Ignore all demo submissions on ICDS

mkangia opened the pull request:
https://dimagi-dev.atlassian.net/browse/ICDS-611
This is a redone of https://github.com/dimagi/commcare-hq/pull/25615 with better approach and commits

##### SUMMARY
We are seeing form submissions in HQ by practice mobile worker on ICDS. This is happening for older versions of CommCare and we have not been able to figure out why that is happening so as a fix,
Ignore all form submissions by a practice mobile worker for ICDS environment.

Best review commit wise

mkangia committed:
add missing tests for demo/normal mode

mkangia committed:
ignore all demo user forms

mkangia committed:
fix documentation

mkangia committed:
temporarily enable it for staging

snopoke commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -20,11 +20,9 @@
-class SubmissionTest(TestCase):
-    maxDiff = None
-
+class BaseSubmissionTest(TestCase):
```
why do these need to be different test classes?

mkangia replied:
Just to keep their intent separate. Initially i put all new 6 tests in `SubmissionTest` but then i thought it would be better to keep them in two different classes based on what they were testing

snopoke commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -20,11 +20,9 @@
-    maxDiff = None
-
+class BaseSubmissionTest(TestCase):
     def setUp(self):
```
almost all of setup and teardown could be done at the class level (except perhpas for deleting the forms and cases in teardown). This would make the tests much faster.

mkangia replied:
Tried that and all tests started failing. Must have done something incorrectly but not up for some test debugging right now so going to leave this as-is

mkangia committed:
fix name[skip ci]

Commit:
Use userID on form instead

The request from demo user or practice mobile worker is made by noauth and hence there is no couch user on the request

Commit:
stickler [skip ci]

Commit:
use noauth requests to replicate expected requests and realign tests

mkangia commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -118,6 +122,64 @@ def test_submit_deprecated_form(self):
+        self.url = self.url + '?submit_mode=demo'
+
+    def test_form_with_demo_user_id_in_demo_mode(self):
+        response = self._submit('demo_mode_simple_form.xml', url=self.url)
```
oh yes, [updated](https://github.com/dimagi/commcare-hq/pull/25711/commits/e2110a8d61293a230d50f6ed0354d87852bea4cc).
Thanks

mkangia committed:
remove staging env

mkangia commented:
tests run locally once to confirm the change made to remove staging env didn't break anything.

snopoke commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -118,6 +122,91 @@ def test_submit_deprecated_form(self):
+class NoAuthSubmissionTest(BaseSubmissionTest):
+    def setUp(self):
+        super(NoAuthSubmissionTest, self).setUp()
+        self.url = self.url + '?authtype=noauth'
```
rather just don't auth the client: https://github.com/dimagi/commcare-hq/blob/74adc998673b642f1b8573493ea32af4c6f04355/corehq/apps/receiverwrapper/tests/test_submissions.py#L31

mkangia replied:
👍

mkangia replied:
the param however is necessary to replicate the method calls that actually happen, and the authorisation is not used, will remove it nonetheless

snopoke replied:
ok

snopoke commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -118,6 +122,91 @@ def test_submit_deprecated_form(self):
+    """
+    def setUp(self):
+        super(PracticeMobileWorkerSubmissionTest, self).setUp()
+        self.url = self.url + '?submit_mode=demo&authtype=noauth'
```
I think submissions from practice users are authenticate. @shubham1g5 can you confirm?

mkangia replied:
@snopoke Shubham shared that practice user also submits via `noath`

snopoke replied:
ok

Commit:
remove any authentication for demo/practice noauth submissions

snopoke commented on corehq/apps/receiverwrapper/tests/test_submissions.py:
```diff
@@ -118,6 +122,97 @@ def test_submit_deprecated_form(self):
+    def setUp(self):
+        super(PracticeMobileWorkerSubmissionTest, self).setUp()
+        self.url = self.url + '?submit_mode=demo&authtype=noauth'
+        self.domain.secure_submissions = True
```
I don't think this is necessary since you're patching `domain_requires_auth`

mkangia replied:
Ah yes, forgot to remove that after adding the patch, thanks

Commit:
remove unnecessary update since domain_requires_auth is patched

mkangia commented:
verified the tests locally for changes in last [commit](https://github.com/dimagi/commcare-hq/pull/25711/commits/6a2a0d0783514faaa33faa6bdf4bf98dfd7324fa) which should not effect other tests, to merge and deploy. So build failure in the last commit should be okay to ignore which is because the PR was merged.
